kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/v2/engine/EngineProvider.scala

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.ha.v2.engine
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.curator.framework.CuratorFramework
+import org.apache.curator.utils.ZKPaths
+
+import org.apache.kyuubi.Logging
+import org.apache.kyuubi.ha.v2.{InstanceProvider, InstanceProviderBuilder, ProviderFilter}
+import org.apache.kyuubi.ha.v2.engine.filters.{TagProviderFilter, VersionProviderFilter}
+
+class EngineProvider(namespace: String, zkClient: CuratorFramework)
+  extends InstanceProvider[EngineInstance] with Logging {
+
+  val filters: ArrayBuffer[ProviderFilter[EngineInstance]] = ArrayBuffer()
+
+  def addFilter(filter: ProviderFilter[EngineInstance]): Unit = {
+    filters += filter
+  }
+
+  override def getInstances(): Seq[EngineInstance] = {
+    val enginePaths = zkClient.getChildren().forPath(namespace)
+    enginePaths.asScala
+      .map(getEngineInstance(_))
```

Comment:
we don't try to get as much as available engine instances as we can but fail directly if any error occurs

Reply:
Sorry, I don't quite understand this reply. The `EngineProvider.getInstances` method is to get all available engines of a shared user, and then get the most appropriate engine through `ProviderStrategy`.

Reply:
you mention in getEngineInstance that it will throw exceptions when can not parse the data to an engine instance